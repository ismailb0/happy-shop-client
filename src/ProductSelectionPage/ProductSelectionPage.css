.menu {
  height: 100%;
  position: fixed;
  width: 25%;
}

.product-selection-showcase {
  height: 100%;
  left: 25%;
  position: absolute;
  width: 75%;
}
