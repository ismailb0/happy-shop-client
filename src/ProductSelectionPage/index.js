export { default } from './ProductSelectionPage.js';
