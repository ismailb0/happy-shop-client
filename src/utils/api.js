import URI from "urijs";


export const getProducts = (
  start_price,
  end_price,
  category,
  subcategory,
  subsubcategory,
  pagination,
  page,
  results_per_page,
  sorted
) => {
  const params = {};

  if (start_price) {
    params.start_price = start_price
  }
  if (end_price) {
    params.end_price = end_price
  }
  if (category) {
    params.category = category
  }
  if (subcategory) {
    params.subcategory = subcategory
  }
  if (subsubcategory) {
    params.subsubcategory = subsubcategory
  }
  if (pagination) {
    params.pagination = pagination
  }
  if (page) {
    params.page = page
  }
  if (results_per_page) {
    params.results_per_page = results_per_page
  }
  if (sorted) {
    params.sorted = sorted
  }

  const url = new URI('https://happy-shop-isb-api.herokuapp.com/application/products');

  Object.keys(params).forEach(key => url.addQuery(key, params[key]));

  return fetch(url)
    .then(response => response.json())
    .then(products => {
      return products
    }
    )
};
