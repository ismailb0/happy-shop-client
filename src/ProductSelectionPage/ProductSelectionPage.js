import React, { Component } from 'react';
import { getProducts } from "../utils/api";

import './ProductSelectionPage.css';
import Menu from '../Menu';
import Showcase from '../Showcase';


class ProductSelectionPage extends Component {

  state = {
    startPrice: null,
    endPrice: null,
    category: 'Makeup',
    subcategory: null,
    subsubcategory: null,
    pagination: true,
    page: 1,
    resultsPerPage: 10,
    sortedBy: "best_selling",
    products: [],
    priceRanges: [],
    count: 0
  }

  componentDidMount() {
    this.fetchData(this.state);
  }

  handleSortChange = (sortedBy) => {
    this.setState({
      sorted: sortedBy
    });
    this.fetchData({
      ...this.state,
      sortedBy
    });
  }

  handleCategoryChange = (category) => {
    this.setState({
      category,
      subcategory: null,
      subsubcategory: null,
    });
    this.fetchData({
      ...this.state,
      category,
      subcategory: null,
      subsubcategory: null,
    });
  }

  handleSubCategoryChange = (subcategory) => {
    this.setState({
      category: null,
      subcategory,
      subsubcategory: null,
    });
    this.fetchData({
      ...this.state,
      category: null,
      subcategory,
      subsubcategory: null,
    });
  }

  handleSubSubCategoryChange = (subsubcategory) => {
    this.setState({
      category: null,
      subcategory: null,
      subsubcategory,
    });
    this.fetchData({
      ...this.state,
      category: null,
      subcategory: null,
      subsubcategory,
    });
  }

  handleResultsPerPageChange = (resultsPerPage) => {
    this.setState({ resultsPerPage });
    this.fetchData({
      ...this.state,
      resultsPerPage
    });
  }

  fetchData = (parameters) => {
    getProducts(
      parameters.startPrice,
      parameters.endPrice,
      parameters.category,
      parameters.subcategory,
      parameters.subsubcategory,
      parameters.pagination,
      parameters.page,
      parameters.resultsPerPage,
      parameters.sortedBy
    )
    .then(response => {
      const products = response.products
      const count = response.count
      if (products) {
        let filteredProducts = products
        if (this.state.startPrice !== null && this.state.endPrice !== null) {
          filteredProducts = filteredProducts.filter((product) => {
            let shouldProductBeKept = false;
            this.state.priceRanges.forEach((priceRange) => {
              if (product.price >= priceRange.startPrice && product.price <= priceRange.endPrice) {
                shouldProductBeKept = true
              }
            })
            return shouldProductBeKept
          })
        }

        this.setState({ products: filteredProducts, count });
      }
    })
    .catch(error => {
      console.error(error);
      this.setErrorMessage("Failed to get logs data.");
    })
  };

  handlePriceRangeChange = (startPrice, endPrice, priceRanges) => {
    this.setState({ startPrice, endPrice, priceRanges });
    this.fetchData({
      ...this.state,
      startPrice,
      endPrice,
    });
  }

  render() {
    return (
      <div>
        <div className="menu">
          <Menu
            handlePriceRangeChange={this.handlePriceRangeChange}
            handleCategoryChange={this.handleCategoryChange}
            handleSubCategoryChange={this.handleSubCategoryChange}
            handleSubSubCategoryChange={this.handleSubSubCategoryChange}
          />
        </div>
        <div className="product-selection-showcase">
          <Showcase
            products={this.state.products}
            handleResultsPerPageChange={this.handleResultsPerPageChange}
            handleSortChange={this.handleSortChange}
          />
        </div>
      </div>
    );
  }
}

export default ProductSelectionPage;
