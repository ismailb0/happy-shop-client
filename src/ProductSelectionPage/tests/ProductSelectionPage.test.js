import React from 'react';
import { shallow } from "enzyme";
import ProductSelectionPage from '../ProductSelectionPage';
import * as Api from '../../utils/api';

const ReactTestRenderer = require('react-test-renderer');

describe('<ProductSelectionPage />', () => {

  it('handleSortChange set sortedBy in the state and call fetch data', () => {
    const renderedComponent = shallow(<ProductSelectionPage />);
    const componentInstance = renderedComponent.instance()
    jest.spyOn(componentInstance, 'fetchData');
    componentInstance.state.sortedBy = 'best_selling'
    componentInstance.handleSortChange('price')
    expect(componentInstance.state.sortedBy).toEqual('price')
    expect(componentInstance.fetchData).toHaveBeenCalledWith({
      ...componentInstance.state,
      sortedBy: 'price'
    })
  });

  it('handleCategoryChange set category in the state and call fetch data', () => {
    const renderedComponent = shallow(<ProductSelectionPage />);
    const componentInstance = renderedComponent.instance()
    jest.spyOn(componentInstance, 'fetchData');
    componentInstance.state.category = 'Perfume'
    componentInstance.handleCategoryChange('Makeup')
    expect(componentInstance.state.category).toEqual('Makeup')
    expect(componentInstance.fetchData).toHaveBeenCalledWith({
      ...componentInstance.state,
      category: 'Makeup',
      subcategory: null,
      subsubcategory: null,
    })
  });

  it('handleSubCategoryChange set subcategory in the state and call fetch data', () => {
    const renderedComponent = shallow(<ProductSelectionPage />);
    const componentInstance = renderedComponent.instance()
    jest.spyOn(componentInstance, 'fetchData');
    componentInstance.state.subcategory = 'Lips'
    componentInstance.handleSubCategoryChange('Face')
    expect(componentInstance.state.subcategory).toEqual('Face')
    expect(componentInstance.fetchData).toHaveBeenCalledWith({
      ...componentInstance.state,
      category: null,
      subcategory: 'Face',
      subsubcategory: null,
    })
  });

  it('handleSubSubCategoryChange set subsubcategory in the state and call fetch data', () => {
    const renderedComponent = shallow(<ProductSelectionPage />);
    const componentInstance = renderedComponent.instance()
    jest.spyOn(componentInstance, 'fetchData');
    componentInstance.state.subsubcategory = 'Eyebrows'
    componentInstance.handleSubSubCategoryChange('Nail Art')
    expect(componentInstance.state.subsubcategory).toEqual('Nail Art')
    expect(componentInstance.fetchData).toHaveBeenCalledWith({
      ...componentInstance.state,
      category: null,
      subcategory: null,
      subsubcategory: 'Nail Art',
    })
  });

  it('handleResultsPerPageChange set resultsPerPage in the state and call fetch data', () => {
    const renderedComponent = shallow(<ProductSelectionPage />);
    const componentInstance = renderedComponent.instance()
    jest.spyOn(componentInstance, 'fetchData');
    componentInstance.state.resultsPerPage = 12
    componentInstance.handleResultsPerPageChange(10)
    expect(componentInstance.state.resultsPerPage).toEqual(10)
    expect(componentInstance.fetchData).toHaveBeenCalledWith({
      ...componentInstance.state,
      resultsPerPage: 10
    })
  });

  it('handlePriceRangeChange set startPrice and endPrice in the state and call fetch data', () => {
    const renderedComponent = shallow(<ProductSelectionPage />);
    const componentInstance = renderedComponent.instance()
    jest.spyOn(componentInstance, 'fetchData');
    componentInstance.state.startPrice = 0
    componentInstance.state.endPrice = 2
    componentInstance.state.priceRanges = []
    const priceRanges = [{
        startPrice: 10,
        endPrice: 15
      },
      {
        startPrice: 20,
        endPrice: 1000
      }
    ]
    componentInstance.handlePriceRangeChange(10, 1000, priceRanges)
    expect(componentInstance.state.startPrice).toEqual(10)
    expect(componentInstance.state.endPrice).toEqual(1000)
    expect(componentInstance.state.priceRanges).toEqual(priceRanges)
    expect(componentInstance.fetchData).toHaveBeenCalledWith({
      ...componentInstance.state,
      startPrice: 10,
      endPrice: 1000,
    })
  });

  it('handlePageChange set the page in the state and call fetch data', () => {
    const renderedComponent = shallow(<ProductSelectionPage />);
    const componentInstance = renderedComponent.instance()
    jest.spyOn(componentInstance, 'fetchData');
    componentInstance.state.page = 1
    componentInstance.handlePageChange(3)
    expect(componentInstance.state.page).toEqual(3)
    expect(componentInstance.fetchData).toHaveBeenCalledWith({
      ...componentInstance.state,
      page: 3
    })
  });

  it('fetchData should call getProducts with the right parameters', () => {

    const mockSuccessResponse = {};
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });
    jest.spyOn(Api, 'getProducts').mockImplementation(() => mockFetchPromise);

    const renderedComponent = shallow(<ProductSelectionPage />);
    const componentInstance = renderedComponent.instance()

    expect(Api.getProducts).toHaveBeenCalledTimes(1);
  });

  it('should match snapshot for ProductSelectionPage', () => {

    const component = ReactTestRenderer.create(
      <ProductSelectionPage />
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
