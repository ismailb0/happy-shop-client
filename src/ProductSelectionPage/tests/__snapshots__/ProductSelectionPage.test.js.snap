// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductSelectionPage /> should match snapshot for ProductSelectionPage 1`] = `
<div>
  <div
    className="menu"
  >
    <div
      className="menu-item-list"
    >
      <div>
        <div
          className="MuiList-root-1 MuiList-padding-2"
        >
          <div
            className="MuiButtonBase-root-17 MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-13 MuiListItem-button-14"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex="0"
          >
            <div
              className="MuiListItemText-root-20 MuiListItemText-inset-21"
            >
              <div>
                Makeup
              </div>
            </div>
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-26"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
            <span
              className="MuiTouchRipple-root-142"
            />
          </div>
        </div>
      </div>
      <div
        className="MuiList-root-1 MuiList-padding-2"
      >
        <div
          className="MuiButtonBase-root-17 MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-13 MuiListItem-button-14"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex="0"
        >
          <div
            className="MuiListItemText-root-20 MuiListItemText-inset-21"
          >
            <span
              className="MuiTypography-root-40 MuiTypography-subheading-47 MuiListItemText-primary-23"
            >
              <div
                className="menu-item-text"
              >
                Price
              </div>
            </span>
          </div>
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-26"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
          <span
            className="MuiTouchRipple-root-142"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="product-selection-showcase"
  >
    <div
      className="showcase"
    >
      <div
        className="showcase-navigation"
      >
        <form
          autoComplete="off"
          className="form"
        >
          <div
            className="MuiFormControl-root-77"
          >
            <div
              className="label-dropdown-wrapper"
            >
              <div
                className="label"
              >
                Sort By :
              </div>
              <label
                className="MuiFormLabel-root-92 MuiFormLabel-filled-96 MuiInputLabel-root-81 MuiInputLabel-formControl-86 MuiInputLabel-animated-89 MuiInputLabel-shrink-88 MuiInputLabel-outlined-91"
                data-shrink={true}
                htmlFor="outlined-age-simple"
              />
              <div
                className="MuiInputBase-root-119 MuiOutlinedInput-root-106 MuiInputBase-formControl-120 Dropdown-root-76"
                onClick={[Function]}
              >
                <fieldset
                  aria-hidden={true}
                  className="MuiPrivateNotchedOutline-root-136 MuiOutlinedInput-notchedOutline-113"
                  style={
                    Object {
                      "paddingLeft": 8,
                    }
                  }
                >
                  <legend
                    className="MuiPrivateNotchedOutline-legend-137"
                    style={
                      Object {
                        "width": 0,
                      }
                    }
                  >
                    <span
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "&#8203;",
                        }
                      }
                    />
                  </legend>
                </fieldset>
                <div
                  className="MuiSelect-root-99"
                >
                  <div
                    aria-haspopup="true"
                    aria-pressed="false"
                    className="MuiSelect-select-100 MuiSelect-selectMenu-103 MuiSelect-outlined-102 MuiInputBase-input-129 MuiOutlinedInput-input-114"
                    id="select-value"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    role="button"
                    tabIndex={0}
                  >
                    Best Selling
                  </div>
                  <input
                    id="outlined-age-simple"
                    name="value"
                    type="hidden"
                    value="best_selling"
                  />
                  <svg
                    aria-hidden="true"
                    className="MuiSvgIcon-root-26 MuiSelect-icon-105"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </form>
        <form
          autoComplete="off"
          className="form"
        >
          <div
            className="MuiFormControl-root-77"
          >
            <div
              className="label-dropdown-wrapper"
            >
              <div
                className="label"
              >
                View :
              </div>
              <label
                className="MuiFormLabel-root-92 MuiFormLabel-filled-96 MuiInputLabel-root-81 MuiInputLabel-formControl-86 MuiInputLabel-animated-89 MuiInputLabel-shrink-88 MuiInputLabel-outlined-91"
                data-shrink={true}
                htmlFor="outlined-age-simple"
              />
              <div
                className="MuiInputBase-root-119 MuiOutlinedInput-root-106 MuiInputBase-formControl-120 Dropdown-root-76"
                onClick={[Function]}
              >
                <fieldset
                  aria-hidden={true}
                  className="MuiPrivateNotchedOutline-root-136 MuiOutlinedInput-notchedOutline-113"
                  style={
                    Object {
                      "paddingLeft": 8,
                    }
                  }
                >
                  <legend
                    className="MuiPrivateNotchedOutline-legend-137"
                    style={
                      Object {
                        "width": 0,
                      }
                    }
                  >
                    <span
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "&#8203;",
                        }
                      }
                    />
                  </legend>
                </fieldset>
                <div
                  className="MuiSelect-root-99"
                >
                  <div
                    aria-haspopup="true"
                    aria-pressed="false"
                    className="MuiSelect-select-100 MuiSelect-selectMenu-103 MuiSelect-outlined-102 MuiInputBase-input-129 MuiOutlinedInput-input-114"
                    id="select-value"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    role="button"
                    tabIndex={0}
                  >
                    10
                  </div>
                  <input
                    id="outlined-age-simple"
                    name="value"
                    type="hidden"
                    value={10}
                  />
                  <svg
                    aria-hidden="true"
                    className="MuiSvgIcon-root-26 MuiSelect-icon-105"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </form>
        <div
          className="pagination"
        />
      </div>
      <div
        className="product-showcase-wrapper"
      >
        <div
          className="product-showcase"
        />
      </div>
    </div>
  </div>
</div>
`;
